src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerGenotypingEngine.java

```diff
@@ -211,7 +221,32 @@ public CalledHaplotypes assignGenotypeLikelihoods(final List<Haplotype> haplotyp
                     logger.info("Genotyping event at " + loc + " with alleles = " + mergedVC.getAlleles());
                 }
 
-                ReadLikelihoods<Allele> readAlleleLikelihoods = readLikelihoods.marginalize(alleleMapper, new SimpleInterval(mergedVC).expandWithinContig(ALLELE_EXTENSION, header.getSequenceDictionary()));
+                // if the number of allele is so high that enumerating all possible genotypes is impractical,
+                // as determined by MAX_GENOTYPE_COUNT_TO_ENUMERATE,
+                // trim alleles that are not well supported by good-scored haplotypes,
+                // otherwise alias, i.e. no trimming if genotype count is small
+                final int originalAlleleCount = alleleMapper.size();
+                Integer practicalAlleleCount = practicaAlleleCountForPloidy.get(ploidy);
+                if(practicalAlleleCount==null) { // maximum allele count given this ploidy and MAX_GENOTYPE_COUNT_TO_ENUMERATE hasn't been computed
+                    practicalAlleleCount = GenotypeLikelihoodCalculators.computeMaxAcceptableAlleleCount(ploidy, MAX_GENOTYPE_COUNT_TO_ENUMERATE);
+                    practicaAlleleCountForPloidy.put(ploidy, practicalAlleleCount);
+                }
+
+                Map<Allele, List<Haplotype>> practicalAlleleMapper = null;
+                if (practicalAlleleCount < originalAlleleCount) {
+                    practicalAlleleMapper = reduceNumberOfAlternativeAllelesBasedOnHaplotypesScores(alleleMapper, practicalAlleleCount);
+                    if( configuration.DEBUG && logger != null ) {
+                        logger.warn(String.format("Removed alt alleles where ploidy is %d and original allele count is %d, whereas after trimming the allele count becomes %d",
+                                ploidy, originalAlleleCount, practicalAlleleCount));
+                        logger.warn(String.format("Alleles kept are:%s", practicalAlleleMapper.keySet()));
+                    }
+                }else{
+                    practicalAlleleMapper = alleleMapper;
+                }
+
+                ReadLikelihoods<Allele> readAlleleLikelihoods = readLikelihoods.marginalize(practicalAlleleMapper,
+                                                                                            new SimpleInterval(mergedVC).expandWithinContig(ALLELE_EXTENSION, header.getSequenceDictionary())
+                                                                                            );
```

Comment:
If you want to split this up I would still keep the `);` on the second line I think.